feat(arithmetic): Do validation on column edit

wmak opened the pull request:
- This shows arithmetic validation errors in the column editor so users
  don't have to apply changes before being able to find out what's wrong
- The syntax errors are a bit rough right now, but I think this is a
  decent start for using pegjs

# Preview
<img width="934" alt="image" src="https://user-images.githubusercontent.com/4205004/126575257-927d553c-523b-425e-8085-10f7eaa8ae51.png">

wmak committed:
feat(arithmetic): Do validation on column edit

- This shows arithmetic validation errors in the column editor so users
  don't have to apply changes before being able to find out what's wrong
- The syntax errors are a bit rough right now, but I think this is a
  decent start for using pegjs

Zylphrex commented on static/app/components/arithmeticInput/parser.tsx:
```diff
@@ -1,5 +1,10 @@
+
 import grammar from './grammar.pegjs';
 
+const MAX_OPERATORS = 10;
```
Can we add a comment that this should stay in sync with the constant in the backend?

Zylphrex commented on static/app/components/arithmeticInput/parser.tsx:
```diff
@@ -49,19 +72,20 @@ function flatten(remaining: Array<Operation>): Operation | undefined {
-};
-
-export function parseArithmetic(query: string): Operation | null {
+export function parseArithmetic(query: string): {result: term; error: string} {
```
Nit: `term` -> `Term`?

Zylphrex commented on static/app/views/eventsV2/table/columnEditCollection.tsx:
```diff
@@ -36,6 +37,7 @@ type State = {
   draggingIndex: undefined | number;
   draggingTargetIndex: undefined | number;
   draggingGrabbedOffset: undefined | {x: number; y: number};
+  error: Map<number, undefined | string>;
```
Is `undefined` necessary here?

Zylphrex commented on static/app/views/eventsV2/table/columnEditCollection.tsx:
```diff
@@ -120,6 +123,16 @@ class ColumnEditCollection extends React.Component<Props, State> {
 
   handleUpdateColumn = (index: number, column: Column) => {
     const newColumns = [...this.props.columns];
+    if (column.kind === 'equation') {
+      this.setState(prevState => {
+        const error = prevState.error;
+        error[index] = parseArithmetic(column.field).error;
+        return {
+          ...prevState,
+          error,
+        };
+      });
+    }
```
Does `error` need to be cleared/reset as new columns are added, removed, reordered, etc.?

wmak committed:
ref: Addressing PR comments

Zylphrex commented on static/app/views/eventsV2/table/columnEditCollection.tsx:
```diff
@@ -120,11 +138,25 @@ class ColumnEditCollection extends React.Component<Props, State> {
   handleUpdateColumn = (index: number, column: Column) => {
     const newColumns = [...this.props.columns];
+    if (column.kind === 'equation') {
+      this.setState(prevState => {
+        const error = prevState.error;
+        error.set(index, parseArithmetic(column.field).error);
+        return {
+          ...prevState,
+          error,
+        };
+      });
```
This feels like incorrect react as you're mutating the same map when setting the state. I believe the only reason why this is working is because the `onChange` below is causing the props to change which is causing the re-render.

wmak replied:
Hmm even with the onChange commented out the error state still updates

dashed replied:
I think you might need to clone the map. i.e. `newErrors = new Map(error)`

wmak replied:
🙏 sweet updated

Zylphrex commented on tests/js/spec/views/eventsV2/table/columnEditModal.spec.js:
```diff
@@ -408,6 +408,33 @@ describe('EventsV2 -> ColumnEditModal', function () {
         newWrapper.find('RowContainer button[aria-label="Drag to reorder"]')
       ).toHaveLength(0);
     });
+    it('handles equations being deleted', function () {
```
And a test for the following?
1. the error message has changed
2. the order has changed

wmak committed:
ref: Addressing PR comments

wmak committed:
ref: cloning the error map to avoid state issues

Zylphrex commented on static/app/components/arithmeticInput/parser.tsx:
```diff
@@ -49,19 +74,20 @@ function flatten(remaining: Array<Operation>): Operation | undefined {
-  } catch (e) {
-    // TODO(wmak): Should we capture these errors somewhere?
+    const result = grammar.parse(query, {tc});
+    return {result, error: tc.errors[0]};
```
`tc.errors[0]` can be `undefined` if there are no errors. This would mean the types in `columnEditCollection` should either be `Map<number, string | undefined>` or `Map<number, string | null>` so we're explicit about the no error case.

wmak committed:
ref: fixing type on parseArithmetic, rechecking errors

- Basically gave up on trying to shuffle items in a map, going to just
  call checkColumnErrors again on update/delete instead
